File: go/pricing/applieddiscount_test.go
// ABOUTME: Unit tests for the AppliedDiscount domain type.
// ABOUTME: Verifies discount creation with and without allocations.
package pricing

import (
	"testing"

	"github.com/stretchr/testify/assert"
	"promotionengine/domain"
)

func TestNewAppliedDiscount(t *testing.T) {
	promoID := domain.MustPromotionID("PROMO10")
	amount := domain.Euros(5.00)
	target := "cart"
	details := "10% off cart"

	discount := NewAppliedDiscount(promoID, amount, target, details)

	assert.Equal(t, promoID, discount.PromotionID)
	assert.Equal(t, amount, discount.Amount)
	assert.Equal(t, target, discount.Target)
	assert.Equal(t, details, discount.Details)
	assert.Nil(t, discount.Allocations)
}

func TestNewAppliedDiscountWithAllocations(t *testing.T) {
	promoID := domain.MustPromotionID("PROMO10")
	amount := domain.Euros(5.00)
	target := "line"
	details := "10% off per line"

	appleSku := domain.MustSku("APPLE")
	bananaSku := domain.MustSku("BANANA")
	allocations := map[domain.Sku]domain.Money{
		appleSku:  domain.Euros(3.00),
		bananaSku: domain.Euros(2.00),
	}

	discount := NewAppliedDiscountWithAllocations(promoID, amount, target, details, allocations)

	assert.Equal(t, promoID, discount.PromotionID)
	assert.Equal(t, amount, discount.Amount)
	assert.Equal(t, target, discount.Target)
	assert.Equal(t, details, discount.Details)
	assert.NotNil(t, discount.Allocations)
	assert.Equal(t, domain.Euros(3.00), discount.Allocations[appleSku])
	assert.Equal(t, domain.Euros(2.00), discount.Allocations[bananaSku])
}

File: typescript/src/pricing/AppliedDiscount.test.ts
import { describe, it, expect } from 'vitest';
import { AppliedDiscount } from './AppliedDiscount';
import { PromotionId } from '../domain/PromotionId';
import { Money } from '../domain/Money';
import { Sku } from '../domain/Sku';

describe('AppliedDiscount', () => {
  describe('construction', () => {
    it('should construct without allocations', () => {
      const promotionId = new PromotionId('PROMO-123');
      const amount = Money.euros('10.00');
      const discount = new AppliedDiscount(promotionId, amount, 'cart', 'Test discount');

      expect(discount.promotionId).toBe(promotionId);
      expect(discount.amount).toBe(amount);
      expect(discount.target).toBe('cart');
      expect(discount.details).toBe('Test discount');
      expect(discount.allocations).toBeUndefined();
    });

    it('should construct with allocations', () => {
      const promotionId = new PromotionId('PROMO-123');
      const amount = Money.euros('10.00');
      const skuA = new Sku('SKU-A');
      const skuB = new Sku('SKU-B');
      const allocations = new Map<Sku, Money>([
        [skuA, Money.euros('6.00')],
        [skuB, Money.euros('4.00')],
      ]);

      const discount = new AppliedDiscount(
        promotionId,
        amount,
        'line',
        'Test discount with allocations',
        allocations
      );

      expect(discount.promotionId).toBe(promotionId);
      expect(discount.amount).toBe(amount);
      expect(discount.target).toBe('line');
      expect(discount.details).toBe('Test discount with allocations');
      expect(discount.allocations).toBe(allocations);
      expect(discount.allocations?.size).toBe(2);
    });
  });
});
